# Individual project

#### Dragomir Nedev - M00724882

# Preliminary

## Importing packages

In [118]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score, cross_val_predict, cross_validate
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score, roc_auc_score, precision_score, recall_score, roc_curve, f1_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import mean_squared_error
from sklearn import metrics
from sklearn.preprocessing import StandardScaler
from sklearn.utils import resample
from pprint import pprint
from sklearn.model_selection import RandomizedSearchCV

In [119]:
import warnings
warnings.filterwarnings("ignore")

## Loading the dataset

In [120]:
df = pd.read_csv('synthetic.csv')

In [121]:
X = df.drop('y', axis = 1)
y = pd.DataFrame(df['y'])

## Inspecting the data

In [122]:
df.head()

x1        x2        x3        x4         x5         x6         x7  \
0 -14.698830  2.369710  1.089267 -1.262030 -15.650082 -16.665997  15.909853   
1  -8.457451  2.182712  0.972360 -4.255289 -11.524392  -4.843399   9.557964   
2  -6.541517  1.263892 -0.494469 -2.562072  -8.979410 -23.632245  15.740920   
3 -18.139840  1.569545 -3.286717 -4.255045 -16.146687 -25.893126  12.005963   
4 -12.500957  2.313632  5.227138  2.586718 -15.022213  -3.105726  18.070314   

          x8         x9       x10  ...       x22        x23       x24  \
0 -11.121045  18.275820 -2.405075  ... -5.421817  15.233291 -3.484405   
1 -10.145921   6.655710 -2.821156  ... -5.398857  20.342647 -5.395054   
2  -4.460916 -16.528412 -3.901285  ... -5.339781  10.859401 -2.095555   
3  -2.228017   5.853151 -2.951831  ... -5.652446  -8.674892 -9.665123   
4  -7.745197   0.300133 -3.364458  ... -5.551594  13.195368 -5.089818   

         x25        x26        x27       x28        x29       x30  y  
0   2.755223   9.766386   6.419560  6.618973  15.171849  1.926773  1  
1   2.816668  14.932127   9.134028  4.826775  12.077634  3.397375  0  
2   2.945595  14.778588   2.711564 -0.090958  -5.467509  3.088641  0  
3   8.876766  22.335086  10.194627  2.720710  -1.787331 -0.291131  0  
4  12.362742  22.624796   4.407471  3.022274  -1.705888  6.650217  0  

[5 rows x 31 columns]

In [123]:
X.shape

(1000, 30)

In [124]:
y.shape

(1000, 1)

In [125]:
X.describe()

x1           x2           x3           x4           x5  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean    -13.028746     2.182041    -0.331036    -1.501078   -12.622918   
std       3.659720     1.314388     4.259927     1.922640     3.604514   
min     -25.548066    -1.599455   -14.930338   -10.215498   -24.600418   
25%     -15.588659     1.285855    -3.149624    -2.808884   -15.109200   
50%     -13.072938     2.170483    -0.367062    -1.510223   -12.498793   
75%     -10.534016     3.021294     2.485166    -0.237209   -10.214818   
max      -2.382520     6.026316    14.980421     5.101086     2.182904   

                x6           x7           x8           x9          x10  ...  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000  ...   
mean    -10.854249    15.199978    -1.989472     6.407913    -2.926042  ...   
std       9.750920     7.206344     6.454849    16.872978     0.673362  ...   
min     -55.753091    -4.320908   -22.643235   -51.040173    -4.907236  ...   
25%     -17.120274    10.231755    -6.188742    -4.853568    -3.383162  ...   
50%     -11.170167    15.196222    -2.026093     6.431541    -2.928571  ...   
75%      -4.522221    19.901376     2.392737    18.145285    -2.487445  ...   
max      23.826332    36.646915    19.820630    55.897492    -0.712244  ...   

               x21          x22          x23          x24          x25  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean     -0.909727    -5.472288    10.543841    -6.003123     3.746927   
std       7.484566     0.272104     8.311382     1.873970     4.962534   
min     -24.704022    -6.378320   -14.553686   -12.804169   -10.970233   
25%      -5.849225    -5.666194     4.728117    -7.268277     0.279869   
50%      -0.760839    -5.467538    10.698797    -5.919298     3.841361   
75%       4.087594    -5.287631    16.268073    -4.677299     7.306957   
max      25.311902    -4.671847    36.154495    -0.188857    20.068337   

               x26          x27          x28          x29          x30  
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000  
mean     18.425973     2.742845     3.475475     1.864313    -1.137531  
std       6.134947     7.049830     2.048401    10.351793     8.543692  
min      -1.014732   -18.778590    -2.594584   -30.715194   -27.231646  
25%      14.232877    -2.051034     2.102943    -4.779697    -6.883752  
50%      18.301716     2.989103     3.488600     1.362105    -1.012529  
75%      22.495502     7.393532     4.913292     8.998301     4.753629  
max      36.539176    25.577773     9.590002    31.900767    31.280122  

[8 rows x 30 columns]

In [126]:
df['y'].value_counts()

0    855
1    145
Name: y, dtype: int64

###### We can see that the data is not standardized, however on later stage of development of the model it was discovered insignifficant differences after standardizing, showing that a standardizaiton would not bring value to the model.

###### Another point we could find here is that the data is imbalanced, we have 85.5% "0"s. We could use this as a treshhold value for the model

## Generating train-test split and cross validation

##### We split 80/20%. We have a binary classification problem, even though with imbalanced classes, so it is better to use StratifiedCrossValidation, than just train-test. CV is applied to train data, while test is kept untouched for the final testing. 

In [127]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.8, random_state = 0, stratify = y)

In [128]:
kf = StratifiedKFold(n_splits = 10, shuffle = True, random_state = 0)
kf.get_n_splits(X_train, y_train)

10

## Defining the classifier, feature importances and exploring the parameters in use. Tuning using RandomizedSearchCV

In [129]:
print('Parameters in use:\n')
pprint(clf.get_params())

Parameters in use:

{'bootstrap': True,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 10,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}


In [130]:
n_estimators = [int(x) for x in np.linspace(start = 10, stop = 200, num = 10)]
max_features = [int(x) for x in np.linspace(start = 1, stop = 30, num = 15)]
max_depth = [int(x) for x in np.linspace(1, 11, num = 11)]
min_samples_split = [2, 5, 10]
min_samples_leaf = [1, 2, 4, 6, 8, 10, 12]
bootstrap = [True, False]
#Creating random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}
pprint(random_grid)

{'bootstrap': [True, False],
 'max_depth': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
 'max_features': [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 30],
 'min_samples_leaf': [1, 2, 4, 6, 8, 10, 12],
 'min_samples_split': [2, 5, 10],
 'n_estimators': [10, 31, 52, 73, 94, 115, 136, 157, 178, 200]}


In [131]:
clf_random = RandomizedSearchCV(estimator = clf, param_distributions = random_grid, n_iter = 100, cv = 5, verbose=2, random_state=42)
clf_random.fit(X_train, y_train)

Fitting 5 folds for each of 100 candidates, totalling 500 fits
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True, total=   0.2s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True 


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.1s remaining:    0.0s


[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True, total=   0.2s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True, total=   0.2s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True, total=   0.2s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=2, max_features=1, max_depth=6, bootstrap=True, total=   0.2s
[CV] n_estimators=10, min_samples_split=10, min_samples_leaf=1, max_features=9, max_depth=1, bootstrap=True 
[CV]  n_estimators=10, min_samples_split=10, min_samples_leaf=1, max

[CV]  n_estimators=73, min_samples_split=2, min_samples_leaf=4, max_features=3, max_depth=6, bootstrap=True, total=   0.2s
[CV] n_estimators=73, min_samples_split=2, min_samples_leaf=4, max_features=3, max_depth=6, bootstrap=True 
[CV]  n_estimators=73, min_samples_split=2, min_samples_leaf=4, max_features=3, max_depth=6, bootstrap=True, total=   0.1s
[CV] n_estimators=73, min_samples_split=2, min_samples_leaf=4, max_features=3, max_depth=6, bootstrap=True 
[CV]  n_estimators=73, min_samples_split=2, min_samples_leaf=4, max_features=3, max_depth=6, bootstrap=True, total=   0.1s
[CV] n_estimators=73, min_samples_split=2, min_samples_leaf=4, max_features=3, max_depth=6, bootstrap=True 
[CV]  n_estimators=73, min_samples_split=2, min_samples_leaf=4, max_features=3, max_depth=6, bootstrap=True, total=   0.1s
[CV] n_estimators=10, min_samples_split=10, min_samples_leaf=8, max_features=1, max_depth=3, bootstrap=False 
[CV]  n_estimators=10, min_samples_split=10, min_samples_leaf=8, max_featu

[CV]  n_estimators=115, min_samples_split=2, min_samples_leaf=2, max_features=19, max_depth=10, bootstrap=False, total=   1.5s
[CV] n_estimators=115, min_samples_split=2, min_samples_leaf=2, max_features=19, max_depth=10, bootstrap=False 
[CV]  n_estimators=115, min_samples_split=2, min_samples_leaf=2, max_features=19, max_depth=10, bootstrap=False, total=   1.5s
[CV] n_estimators=115, min_samples_split=2, min_samples_leaf=2, max_features=19, max_depth=10, bootstrap=False 
[CV]  n_estimators=115, min_samples_split=2, min_samples_leaf=2, max_features=19, max_depth=10, bootstrap=False, total=   1.3s
[CV] n_estimators=115, min_samples_split=2, min_samples_leaf=2, max_features=19, max_depth=10, bootstrap=False 
[CV]  n_estimators=115, min_samples_split=2, min_samples_leaf=2, max_features=19, max_depth=10, bootstrap=False, total=   1.4s
[CV] n_estimators=200, min_samples_split=5, min_samples_leaf=8, max_features=17, max_depth=11, bootstrap=False 
[CV]  n_estimators=200, min_samples_split=5,

[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=1, max_features=27, max_depth=1, bootstrap=True, total=   0.5s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=1, max_features=27, max_depth=1, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=1, max_features=27, max_depth=1, bootstrap=True, total=   0.4s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=1, max_features=27, max_depth=1, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=1, max_features=27, max_depth=1, bootstrap=True, total=   0.4s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=1, max_features=27, max_depth=1, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=1, max_features=27, max_depth=1, bootstrap=True, total=   0.4s
[CV] n_estimators=10, min_samples_split=10, min_samples_leaf=8, max_features=23, max_depth=8, bootstrap=False 
[CV]  n_estimators=10, min_samples_split=10, min_samples_le

[CV]  n_estimators=31, min_samples_split=10, min_samples_leaf=2, max_features=1, max_depth=11, bootstrap=True, total=   0.1s
[CV] n_estimators=31, min_samples_split=10, min_samples_leaf=2, max_features=1, max_depth=11, bootstrap=True 
[CV]  n_estimators=31, min_samples_split=10, min_samples_leaf=2, max_features=1, max_depth=11, bootstrap=True, total=   0.1s
[CV] n_estimators=94, min_samples_split=10, min_samples_leaf=2, max_features=19, max_depth=4, bootstrap=True 
[CV]  n_estimators=94, min_samples_split=10, min_samples_leaf=2, max_features=19, max_depth=4, bootstrap=True, total=   0.4s
[CV] n_estimators=94, min_samples_split=10, min_samples_leaf=2, max_features=19, max_depth=4, bootstrap=True 
[CV]  n_estimators=94, min_samples_split=10, min_samples_leaf=2, max_features=19, max_depth=4, bootstrap=True, total=   0.4s
[CV] n_estimators=94, min_samples_split=10, min_samples_leaf=2, max_features=19, max_depth=4, bootstrap=True 
[CV]  n_estimators=94, min_samples_split=10, min_samples_lea

[CV]  n_estimators=31, min_samples_split=5, min_samples_leaf=10, max_features=19, max_depth=3, bootstrap=False, total=   0.2s
[CV] n_estimators=31, min_samples_split=5, min_samples_leaf=10, max_features=19, max_depth=3, bootstrap=False 
[CV]  n_estimators=31, min_samples_split=5, min_samples_leaf=10, max_features=19, max_depth=3, bootstrap=False, total=   0.2s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=10, max_features=17, max_depth=8, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=10, max_features=17, max_depth=8, bootstrap=True, total=   0.8s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=10, max_features=17, max_depth=8, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=10, max_features=17, max_depth=8, bootstrap=True, total=   0.8s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=10, max_features=17, max_depth=8, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_sam

[CV]  n_estimators=31, min_samples_split=5, min_samples_leaf=8, max_features=9, max_depth=11, bootstrap=False, total=   0.2s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=8, max_features=5, max_depth=7, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=8, max_features=5, max_depth=7, bootstrap=True, total=   0.4s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=8, max_features=5, max_depth=7, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=8, max_features=5, max_depth=7, bootstrap=True, total=   0.4s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=8, max_features=5, max_depth=7, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=8, max_features=5, max_depth=7, bootstrap=True, total=   0.4s
[CV] n_estimators=157, min_samples_split=5, min_samples_leaf=8, max_features=5, max_depth=7, bootstrap=True 
[CV]  n_estimators=157, min_samples_split=5, min_samples_leaf=8, ma

[CV]  n_estimators=157, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=10, bootstrap=False, total=   1.7s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=6, max_features=23, max_depth=1, bootstrap=False 
[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=6, max_features=23, max_depth=1, bootstrap=False, total=   0.3s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=6, max_features=23, max_depth=1, bootstrap=False 
[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=6, max_features=23, max_depth=1, bootstrap=False, total=   0.3s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=6, max_features=23, max_depth=1, bootstrap=False 
[CV]  n_estimators=115, min_samples_split=5, min_samples_leaf=6, max_features=23, max_depth=1, bootstrap=False, total=   0.3s
[CV] n_estimators=115, min_samples_split=5, min_samples_leaf=6, max_features=23, max_depth=1, bootstrap=False 
[CV]  n_estimators=115, min_samples_split=5, min_

[CV]  n_estimators=157, min_samples_split=10, min_samples_leaf=8, max_features=11, max_depth=8, bootstrap=True, total=   0.8s
[CV] n_estimators=10, min_samples_split=2, min_samples_leaf=10, max_features=21, max_depth=8, bootstrap=True 
[CV]  n_estimators=10, min_samples_split=2, min_samples_leaf=10, max_features=21, max_depth=8, bootstrap=True, total=   0.1s
[CV] n_estimators=10, min_samples_split=2, min_samples_leaf=10, max_features=21, max_depth=8, bootstrap=True 
[CV]  n_estimators=10, min_samples_split=2, min_samples_leaf=10, max_features=21, max_depth=8, bootstrap=True, total=   0.1s
[CV] n_estimators=10, min_samples_split=2, min_samples_leaf=10, max_features=21, max_depth=8, bootstrap=True 
[CV]  n_estimators=10, min_samples_split=2, min_samples_leaf=10, max_features=21, max_depth=8, bootstrap=True, total=   0.1s
[CV] n_estimators=10, min_samples_split=2, min_samples_leaf=10, max_features=21, max_depth=8, bootstrap=True 
[CV]  n_estimators=10, min_samples_split=2, min_samples_lea

[CV]  n_estimators=94, min_samples_split=2, min_samples_leaf=2, max_features=3, max_depth=2, bootstrap=False, total=   0.2s
[CV] n_estimators=94, min_samples_split=2, min_samples_leaf=12, max_features=7, max_depth=6, bootstrap=False 
[CV]  n_estimators=94, min_samples_split=2, min_samples_leaf=12, max_features=7, max_depth=6, bootstrap=False, total=   0.4s
[CV] n_estimators=94, min_samples_split=2, min_samples_leaf=12, max_features=7, max_depth=6, bootstrap=False 
[CV]  n_estimators=94, min_samples_split=2, min_samples_leaf=12, max_features=7, max_depth=6, bootstrap=False, total=   0.4s
[CV] n_estimators=94, min_samples_split=2, min_samples_leaf=12, max_features=7, max_depth=6, bootstrap=False 
[CV]  n_estimators=94, min_samples_split=2, min_samples_leaf=12, max_features=7, max_depth=6, bootstrap=False, total=   0.4s
[CV] n_estimators=94, min_samples_split=2, min_samples_leaf=12, max_features=7, max_depth=6, bootstrap=False 
[CV]  n_estimators=94, min_samples_split=2, min_samples_leaf=

[CV]  n_estimators=73, min_samples_split=2, min_samples_leaf=10, max_features=13, max_depth=11, bootstrap=False, total=   0.5s
[CV] n_estimators=115, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=1, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=1, bootstrap=True, total=   0.3s
[CV] n_estimators=115, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=1, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=1, bootstrap=True, total=   0.3s
[CV] n_estimators=115, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=1, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=1, bootstrap=True, total=   0.3s
[CV] n_estimators=115, min_samples_split=10, min_samples_leaf=10, max_features=25, max_depth=1, bootstrap=True 
[CV]  n_estimators=115, min_samples_split=10

[CV]  n_estimators=178, min_samples_split=10, min_samples_leaf=6, max_features=19, max_depth=9, bootstrap=False, total=   1.6s
[CV] n_estimators=136, min_samples_split=2, min_samples_leaf=4, max_features=1, max_depth=5, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=2, min_samples_leaf=4, max_features=1, max_depth=5, bootstrap=True, total=   0.2s
[CV] n_estimators=136, min_samples_split=2, min_samples_leaf=4, max_features=1, max_depth=5, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=2, min_samples_leaf=4, max_features=1, max_depth=5, bootstrap=True, total=   0.2s
[CV] n_estimators=136, min_samples_split=2, min_samples_leaf=4, max_features=1, max_depth=5, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=2, min_samples_leaf=4, max_features=1, max_depth=5, bootstrap=True, total=   0.2s
[CV] n_estimators=136, min_samples_split=2, min_samples_leaf=4, max_features=1, max_depth=5, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=2, min_samples_leaf=4, 

[CV]  n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True, total=   0.2s
[CV] n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True 
[CV]  n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True, total=   0.2s
[CV] n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True 
[CV]  n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True, total=   0.2s
[CV] n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True 
[CV]  n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True, total=   0.2s
[CV] n_estimators=52, min_samples_split=10, min_samples_leaf=12, max_features=19, max_depth=3, bootstrap=True 
[CV]  n_estimators=52, min_samples_split=10, min_sam

[CV]  n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True, total=   1.0s
[CV] n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True, total=   0.9s
[CV] n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True, total=   0.8s
[CV] n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True, total=   0.8s
[CV] n_estimators=136, min_samples_split=10, min_samples_leaf=10, max_features=19, max_depth=9, bootstrap=True 
[CV]  n_estimators=136, min_samples_split=10

[Parallel(n_jobs=1)]: Done 500 out of 500 | elapsed:  4.6min finished


RandomizedSearchCV(cv=5, error_score='raise-deprecating',
                   estimator=RandomForestClassifier(bootstrap=True,
                                                    class_weight=None,
                                                    criterion='gini',
                                                    max_depth=None,
                                                    max_features='auto',
                                                    max_leaf_nodes=None,
                                                    min_impurity_decrease=0.0,
                                                    min_impurity_split=None,
                                                    min_samples_leaf=1,
                                                    min_samples_split=2,
                                                    min_weight_fraction_leaf=0.0,
                                                    n_estimators=10,
                                                    n_jobs=None,
  

In [132]:
clf_random.best_params_

{'n_estimators': 200,
 'min_samples_split': 5,
 'min_samples_leaf': 8,
 'max_features': 17,
 'max_depth': 11,
 'bootstrap': False}

In [133]:
y_pred = clf_random.predict(X_test)
print('Unique values: ', np.unique(y_pred))
print('Accuracy score for CV:', accuracy_score(y_test,y_pred).ravel())
prob_pred_y = clf_random.predict_proba(X_test)
prob_pred_y = [p[1] for p in prob_pred_y]
print('ROC AUC Curve:',roc_auc_score(y_test, prob_pred_y).ravel())
print('Classification report')
print(classification_report(y_test,y_pred))

Unique values:  [0 1]
Accuracy score for CV: [0.88]
ROC AUC Curve: [0.85420448]
Classification report
              precision    recall  f1-score   support

           0       0.89      0.98      0.93       171
           1       0.73      0.28      0.40        29

    accuracy                           0.88       200
   macro avg       0.81      0.63      0.67       200
weighted avg       0.87      0.88      0.86       200



In [110]:
#Base model

In [117]:
clf_base = RandomForestClassifier(n_estimators = 42, random_state = 0)
clf_base.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=42,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [109]:
y_pred_base = clf_base.predict(X_test)
print('Unique values: ', np.unique(y_pred_base))
print('Accuracy score for CV:', accuracy_score(y_test,y_pred_base).ravel())
prob_pred_y_base = clf_random.predict_proba(X_test)
prob_pred_y_base = [p[1] for p in prob_pred_y_base]
print('ROC AUC Curve:',roc_auc_score(y_test, prob_pred_y_base).ravel())
print('Classification report')
print(classification_report(y_test,y_pred_base))

Unique values:  [0]
Accuracy score for CV: [0.855]
ROC AUC Curve: [0.84291188]
Classification report
              precision    recall  f1-score   support

           0       0.85      1.00      0.92       171
           1       0.00      0.00      0.00        29

    accuracy                           0.85       200
   macro avg       0.43      0.50      0.46       200
weighted avg       0.73      0.85      0.79       200



### Model comparison
We can clearly observe that the base model has accuracy at the baseline of 85.5% and providing majority class - '0'. On the tuned model we have improvement of 1.5%, to 88.8% and a ROC AUC Curve improvement of 0.5%. The base model is overfitted with recal of 100% and f1 of 92%, where the tuned model has a better bias-variance trade off and the model does not generally select the majority class all the time.